drivers/dht/dht.c

```diff
@@ -19,192 +20,267 @@
+    if (ztimer_now(ZTIMER_USEC) - start < START_THRESHOLD) {
+        DEBUG_PUTS("[dht] error: response high pulse < START_THRESHOLD");
+        return -ENODEV;
+    }
+    return 0;
 }
 
-/**
- * @brief   Wait until the pin @p pin has level @p expect
- *
- * @param   pin     GPIO pin to wait for
- * @param   expect  Wait until @p pin has this logic level
- * @param   timeout Timeout in µs
- *
- * @retval  0       Success
- * @retval  -1      Timeout occurred before level was reached
- */
-static inline int _wait_for_level(gpio_t pin, bool expect, unsigned timeout)
+static inline void _bit_parse(struct dht_data *arg)
 {
-    while (((gpio_read(pin) > 0) != expect) && timeout) {
-        xtimer_usleep(1);
-        timeout--;
+    int8_t pos = arg->bit_pos++;
+    if (arg->bit) {
+        arg->data[pos / 8] |= (0x80U >> (pos % 8));
     }
+}
 
-    return (timeout > 0) ? 0 : -1;
+static inline void _busy_wait_read(struct dht_data *arg)
+{
+    uint32_t start = ztimer_now(ZTIMER_USEC);
+    while (arg->bit_pos != 40) {
+        _wait_for_level(arg->pin, 1, start);
+        start = ztimer_now(ZTIMER_USEC);
+        _wait_for_level(arg->pin, 0, start);
+        arg->bit = (ztimer_now(ZTIMER_USEC) - start > READ_THRESHOLD) ? 1 : 0;
+        _bit_parse(arg);
+    }
 }
 
-static int _read(uint8_t *dest, gpio_t pin)
+static inline int8_t _validate_checksum(uint8_t *data)
 {
-    DEBUG("[dht] read\n");
-    uint16_t res = 0;
-
-    for (int i = 0; i < 8; i++) {
-        uint32_t start, end;
-        res <<= 1;
-        /* measure the length between the next rising and falling flanks (the
-         * time the pin is high - smoke up :-) */
-        if (_wait_for_level(pin, 1, TIMEOUT)) {
-            return -1;
-        }
-        start = xtimer_now_usec();
+    uint8_t sum = 0;
 
-        if (_wait_for_level(pin, 0, TIMEOUT)) {
-            return -1;
-        }
-        end = xtimer_now_usec();
+    for (uint_fast8_t i = 0; i < 4; i++) {
+        sum += data[i];
+    }
+    if (sum != data[BYTEPOS_CHECKSUM]) {
+        return -EIO;
+    }
+    return 0;
+}
 
-        /* if the high phase was more than 40us, we got a 1 */
-        if ((end - start) > PULSE_WIDTH_THRESHOLD) {
-            res |= 0x0001;
-        }
+static inline int8_t _parse_raw_values(dht_t *dev, uint8_t *data)
+{
+    bool is_negative;
+
+    switch (dev->params.type) {
+    case DHT11:
+        DEBUG_PUTS("[dht] parse raw values with DHT11 data format");
+        dev->last_val.humidity = data[BYTEPOS_HUMIDITY_HIGH] * 10
+                               + data[BYTEPOS_HUMIDITY_LOW];
+        /* MSB for integral temperature byte gives sign, remaining is
+         * abs() of value (beware: this is not two's complement!) */
+        is_negative = data[BYTEPOS_TEMPERATURE_LOW] & 0x80;
+        data[BYTEPOS_TEMPERATURE_LOW] &= ~0x80;
+        dev->last_val.temperature = data[BYTEPOS_TEMPERATURE_HIGH] * 10
+                                  + data[BYTEPOS_TEMPERATURE_LOW];
```

Comment:
Indeed, this seems to match what the machine translation from Chinese to English of the [datasheet of the new DHT11](https://pdf1.alldatasheet.com/datasheet-pdf/view/1132088/ETC2/DHT11.html) states:

![image](https://github.com/RIOT-OS/RIOT/assets/2041729/bb99788f-b8c0-4b26-ae80-0da7de32d038)

Reply:
they seem to rewrite the whole data sheet regularly (this has -24.12°C on page 7)
[AOSONG DHT11 2022-12](http://www.aosong.com/userfiles/files/media/DHT11%E6%B8%A9%E6%B9%BF%E5%BA%A6%E4%BC%A0%E6%84%9F%E5%99%A8%E8%AF%B4%E6%98%8E%E4%B9%A6%EF%BC%88%E4%B8%AD%EF%BC%89%20A0-1208.pdf)

(the english datasheet is on "updating"  http://www.aosong.com/en/products-21.html (web.archive.org says for years))

the temperature low bits are used differently  ( the 2022-12 aosong.com data sheet uses interprets them as 0.01°C per LSB)

While looking into the Datasheet I often found the note ```Note: The host reads temperature and humidity data from DHT11 always previous measurements, such as the two measured time interval is long, please read twice in a row is the second time in real time temperature and humidity values.
```
maybe we should add that to the Documentation